pkg/apiserver/webhooks.go

```diff
@@ -226,29 +238,130 @@ func createValidatingWebhook(ctx context.Context, webhookNs string, caCert []byt
-		TLSConfig: &tls.Config{
-			Certificates: []tls.Certificate{cert},
-		},
+	return cert, nil
+}
+
+// watch for changes on the mount path of the secret as volume
+func watchCertificates(directory, certFile, keyFile string) {
+	//set up a watcher for the cert
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		klog.Fatalf("failed to start certificate watcher: %v", err)
 	}
+	defer watcher.Close()
+
+	// if the watcher notices any changes on the mount dir of the secret such as creations or writes to the files in this dir
+	// attempt to load tls cert using the new cert and key files provided and log output
+	done := make(chan bool)
 	go func() {
-		err = server.ListenAndServeTLS("", "")
-		if err != nil {
-			klog.Errorf("could not start server: %v", err)
+		for {
+			select {
+			case event, ok := <-watcher.Events:
+				if !ok {
+					return
+				}
+				if event.Op&fsnotify.Create == fsnotify.Create || event.Op&fsnotify.Write == fsnotify.Write {
+					_, err := loadCertificate(certFile, keyFile)
+					if err != nil {
+						klog.Errorf("Failed to load updated certificate: %v", err)
+					} else {
+						klog.Info("Certificate reloaded successfully")
+					}
+				}
+			case err, ok := <-watcher.Errors:
+				if !ok {
+					return
+				}
+				klog.Errorf("Watcher error: %v", err)
+			}
 		}
 	}()
-	return err
+	// start the watcher with the dir to watch
+	err = watcher.Add(directory)
+	if err != nil {
+		klog.Fatalf("Error in running watcher: %v", err)
+	}
+
+	<-done
+}
+// return current cert
+func getCertificate(*tls.ClientHelloInfo) (*tls.Certificate, error) {
+	mu.Lock()
+	defer mu.Unlock()
+	return &cert, nil
+}
+
+func runWebhookServer(certStorageDir string, useCertManWebhook bool) error {
+	certFile := filepath.Join(certStorageDir, "tls.crt")
+	keyFile := filepath.Join(certStorageDir, "tls.key")
+	// load the cert for the first time
+
+	if useCertManWebhook {
+		_, err := loadCertificate(certFile, keyFile)
+		if err != nil {
+			klog.Errorf("failed to load certificate: %v", err)
+		}
+		// Start watching the certificate files for changes
+		// watch for changes in directory where secret is mounted
+		go watchCertificates(certStorageDir, certFile, keyFile)
+
+		klog.Infoln("Starting webhook server")
+		http.HandleFunc(serverEndpoint, validateDeletion)
+		server := http.Server{
+			Addr: fmt.Sprintf(":%d", webhookServicePort),
+			TLSConfig: &tls.Config{
+				GetCertificate: getCertificate,
+			},
+		}
+		go func() {
+			err = server.ListenAndServeTLS("", "")
+			if err != nil {
+				klog.Errorf("could not start server: %v", err)
+			}
+		}()
+		return err
+
+	} else {
+		cert, err := tls.LoadX509KeyPair(certFile, keyFile)
+		if err != nil {
+			return err
+		}
+		klog.Infoln("Starting webhook server")
+		http.HandleFunc(serverEndpoint, validateDeletion)
+		server := http.Server{
+			Addr: fmt.Sprintf(":%d", webhookServicePort),
+			TLSConfig: &tls.Config{
+				Certificates: []tls.Certificate{cert},
+			},
+		}
+		go func() {
```

Comment:
Some duplication here we could prob extract

Reply:
- yes i took note of this when i first wrote it. there is too much duplication. i attempted to clean it up but some methods were not playing nicely with the if statement.
- ill dedicate some time to cleaning it up.